Database JUnit Tests

Rsl1122 opened the pull request:
# Ready for Merge
### PR Contents
This is a PR for Database JUnit Tests that use PowerMockito for solving JavaPluginLoader requirement.

**Test System**
Utilities for Mocking essential plugin components for initializing subsystems without enabling the actual plugin.

Current Coverage: 

Class | Methods | Lines | Goal (Lines)
-- | -- | -- | --
AbstractSQLDatabaseManager | 85% | 82% | 85% (Attained)
SQLiteDatabaseManager | 100% | 58%
DatabaseUpdater | 85% | 32%
SQLReadOperation | 100% | 50%
SQLWriteOperation | 66% | 31%

**Basic tests**

Test | Description | Status
-- | -- | --
testInitialization | Initializes & Closes the database |  Passed
testGetAchievementList | Test for getPlayerAchievementsList | Passed
testAchievementCount | Test for getPlayersAchievementsAmount |  Passed
testAchievementDateRegistration | Test for getPlayerAchievementDate | Passed
testPlayerAchievementAmount | Test for getPlayerAchievementsAmount | Passed
testDeleteAchievement | Test for deletePlayerAchievement | Passed
testDeleteAchievementQuotes | testDeleteAchievement with additional quotes | Passed
testConnectionUpdate | Test for updateAndGetConnection & getConnectionsAmount | Passed
testGetTopAchievements | Test for getTopList | Passed
testGetAchievementNameList | Test for getPlayerAchievementNamesList | Passed
testHasAchievement | Test for hasPlayerAchievement | Passed
testGetPlayerConnectionDate | Test for getPlayerConnectionDate | Passed
testClearConnection | Test for clearConnection | Passed

**Null Safety tests**

Test | Description | Status
-- | -- | --
testRegisterNullUUID | Registers achievement with null UUID | IllegalArgumentException: Invalid UUID string: null
testGetMethodsForNullUUIDExceptions | null UUID to db & check for exceptions in get methods | IllegalArgumentException: Invalid UUID string: null
testRegisterNullAch | Registers achievement with null name String | Failed (Achievement in db)
testRegisterNullMsg | Registers achievement with null message String | Failed (Message was null)


*Seems like multiple connections are causing flakiness in the tests. Possibly solved by* 6bd42af93c9881f1ea186a9e67a8fd11ae6ca764

## Possible errors that may occur during tests
### The JBDC driver for the chosen database type was not found.
```
java.sql.SQLException: No suitable driver found for jdbc:sqlite:/home/ristolah/NetBeansProjects/Advanced-Achievements/AdvancedAchievements/target/test-folder/achievements.db
	at java.sql.DriverManager.getConnection(DriverManager.java:689)
	at java.sql.DriverManager.getConnection(DriverManager.java:270)
	at com.hm.achievement.db.SQLiteDatabaseManager.createSQLConnection(SQLiteDatabaseManager.java:66)
	at com.hm.achievement.db.AbstractSQLDatabaseManager.getSQLConnection(AbstractSQLDatabaseManager.java:148)
	at com.hm.achievement.db.AbstractSQLDatabaseManager.initialise(AbstractSQLDatabaseManager.java:91)
``` 
This is usually caused by full Craftbukkit jar (That contains SQLite driver) not being available during test. Should be prevented by f00050daf039658060c0ac3233455e16cbe36455

PyvesB commented:
Hello @Rsl1122 ,

This looks promising, thanks for your work! :wink:

Two preliminary thoughts that come to mind after going through your commits so far:
- I'm guessing you chose to use PowerMock because of final methods in JavaPlugin such as `getLogger`. Mockito 2.1+ has a nice opt-in incubating feature that allows to mock such methods, you can enable it by referring to [this documentation page](https://github.com/mockito/mockito/wiki/What's-new-in-Mockito-2#mock-the-unmockable-opt-in-mocking-of-final-classesmethods). You can therefore probably remove PowerMock and stick with Mockito.
- Could you please move the test files so that they follow the Maven's [Standard Directory Layout](https://maven.apache.org/guides/introduction/introduction-to-the-standard-directory-layout.html)?

Cheers,

Pyves

Rsl1122 commented:
Alright I'll see what I can do :)

Rsl1122 commented:
Should the registerAchievement method throw NullPointerException if one of the arguments is null? (Currently all null tests throw the exception)

PyvesB commented:
Normally the arguments should not be null as the `Name` and `Message` parameters are compulsory in the configuration. Nevertheless, if ever `Name` happened to be null, the method should probably blow up, if `Message` was null, we could possibly persist NULL to the database or an empty string.

Rsl1122 commented:
Alright I'll make the tests reflect that

Rsl1122 commented:
Alright I think the tests are done, I didn't test the NormalAchievements and MultipleAchievements methods. Some lines that deal with special cases (Double quotes & Display name in plugin cache) were also not covered.

The four NullSafety tests don't pass, but that would require changes to the actual database methods.
Awaiting for a review :)

PyvesB commented on src/test/java/com/hm/achievement/db/SQLiteDatabaseBasicTest.java:
```diff
@@ -0,0 +1,228 @@
+
+    @Test
+    public void testInitialization() throws PluginLoadError {
+        initDB();
```
This is a heavy method call. Underneath the hoods we are creating a whole new database file with close to 50 database tables at the beginning of each test. On my old laptop equipped with a hard disk drive, the unit tests take over a minute to complete.

What do you think about moving this to a preceding `@BeforeClass` method or if not possible, avoid systematically deleting the database?  We could then simply clear the `achievements` table at the beginning of each test. This should speed them up quite a bit! 😉

Rsl1122 replied:
When testing on my own plugin at one point `@BeforeClass` was called like `@Before`

I can't remember what caused it so this is still an open option

Rsl1122 replied:
Done, improved test time drastically.

PyvesB replied:
Indeed, the tests are much faster now!

Rsl1122 commented on src/test/java/com/hm/achievement/db/SQLiteDatabaseTest.java:
```diff
@@ -0,0 +1,35 @@
+
+    void sleep25ms() {
+        try {
+            Thread.sleep(25);
```
It might be possible to set the maximum pool size to 1 during the tests

PyvesB replied:
Setting the maximum pool size to 1 won't help here. `SingleThreadExecutor` basically means that there is one worker thread executing tasks one by one in the executor service. Nevertheless this thread is different from the one used to execute the tests, so the database write operations will still happen in parallel.

You have to use an executor service that will execute tasks synchronously on the same thread as the one running the tests. Guava has some simple implementations available, for instance ` MoreExecutors.newDirectExecutorService()`. 😉

Rsl1122 commented:
This first time I'm receiving actual feedback on my tests, so all feedback will definitely help me in the future, thanks!

You brought up some good points and I'll try to improve the PR according to the comments. :)

Rsl1122 commented on src/test/java/utilities/TestFolder.java:
```diff
@@ -0,0 +1,56 @@
+ *
+ * @author Rsl1122
+ */
+public class TestFolder {
```
No I did not. I'm also wondering, now that we're using Mockito instead of Power Mock, if JUnit 5 would be compatible

Rsl1122 replied:
Seems like TemporaryFolder is used per test (method) while TestFolder (mine) is used by mocks (now performed in `@BeforeClass`)

PyvesB replied:
I believe you can use `TemporaryFolder` with a `@ClassRule` so that it is used for the whole class rather than per test method. ^^

Rsl1122 commented on src/test/java/com/hm/achievement/db/SQLiteDatabaseBasicTest.java:
```diff
@@ -0,0 +1,228 @@
+        sleep25ms();
+
+        List<String> achievements = db.getPlayerAchievementsList(testUUID);
+        assertFalse("List was empty", achievements.isEmpty());
```
Yes

Rsl1122 commented on src/test/java/com/hm/achievement/db/SQLiteDatabaseNullSafetyTest.java:
```diff
@@ -0,0 +1,117 @@
+ * @author Rsl1122
+ */
+@RunWith(MockitoJUnitRunner.class)
+public class SQLiteDatabaseNullSafetyTest extends SQLiteDatabaseTest {
```
I think the database itself should at least check for null UUID. Currently if someone manages to drop in a null UUID it breaks some of the get methods due to `UUID.fromString(null)`

PyvesB replied:
After a quick check, the UUIDs used in the plugin internals seem to always be generated by `player.getUniqueId()`, so they cannot be null. The UUID can only be null when the database methods are called from the API layer. Putting the validation there rather than in SQLDatabaseManager allows to squeeze in a tiny bit of performance when the database methods are called by the plugin directly, as we don't need to perform validation on something we know will be correct, and avoids adding complexity to an already complicated class. 😉

Rsl1122 commented on src/test/java/com/hm/achievement/db/SQLiteDatabaseBasicTest.java:
```diff
@@ -0,0 +1,228 @@
+        String actual = achievements.get(0);
+        System.out.println("Saved Achievement: " + actual);
+        assertEquals(testAchievement, actual);
+
```
Removed in commit 2a0d7448d5fec4871593fc3eac05dfdcf485e3c2 :P

Rsl1122 commented:
Requested changes have been made.
- Improved test execution time
- Changes to assertions (about the same still)
- Removed java.util.Date
- Removed some extra empty lilnes

Unfortunately it seems like SingleThreadExecutor didn't make the tests less flaky (Probably made it more flaky) so it was reverted.

PyvesB commented on src/test/java/com/hm/achievement/db/SQLiteDatabaseBasicTest.java:
```diff
@@ -0,0 +1,220 @@
+    }
+
+    @Test
+    public void testDeleteAchievement() throws PluginLoadError {
```
`throws PluginLoadError` no longer needed here.

Rsl1122 replied:
Seems like IntelliJ code inspection missed that one

PyvesB commented:
Looking good! I've responded to some of the pending discussions above.

Rsl1122 commented:
`MoreExecutors.newDirectExecutorService()` was a really good idea!
Now the tests pass without sleeping!
There doesn't seem to be any more flakiness going on either.

I also disabled the Null Safety tests (`@Ignore`) so that you can decide what to do with those later.

PyvesB commented:
Awesome! 

One last detail before I merge. `@ClassRule` will automatically run some methods before and after the tested class. Therefore you don't actually need to call `temporaryFolder.delete()` explicitly, as it's part of the `after` method of the rule. Similarly, you could replace `.mockDataFolder(temporaryFolder.newFolder())` by `.mockDataFolder(temporaryFolder.getRoot())`. 😉

Rsl1122 commented:
Alright, done :)

PyvesB commented:
Merged. Thanks again for your contribution, I'll take care of the _SQLiteDatabaseNullSafetyTest_ class in the coming days!

Rsl1122 commented:
It was great working with you, thanks for all the feedback!

I made a write up in form of a tutorial for others too so that they can get their testing started :)
https://www.spigotmc.org/threads/tutorial-plugin-sqlite-junit-testing.293352/
